File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys


def check_row(a, n):
    for i in range(9):
        if n == sudoku[a][i]:
            return False
    return True


def check_col(a, n):
    for i in range(9):
        if n == sudoku[i][a]:
            return False
    return True


def check_3x3_square(y, x, n):
    for r in range(3):
        for c in range(3):
            if n == sudoku[y - y % 3 + r][x - x % 3 + c]:
                return False
    return True


def solve_sudoku(n):
    if n == len(fill):
        for s in sudoku:
            print(*s)
        exit()
    for i in range(1, 10):
        y, x = fill[n]
        if check_row(y, i) and check_col(x, i) and check_3x3_square(y, x, i):
            sudoku[y][x] = i
            solve_sudoku(n + 1)
            sudoku[y][x] = 0


sudoku = [list(map(int, sys.stdin.readline().split())) for _ in range(9)]
fill = []
for y in range(9):
    for x in range(9):
        if sudoku[y][x] == 0:
            fill.append((y, x))
solve_sudoku(0)

File: 백준/Gold/2580. 스도쿠/스도쿠.cc
#include <iostream>
using namespace std;

typedef pair<int, int> pii;

int grid[9][9];
int idx = 0;
pii pos[81];

bool checkRow(int r, int n) {
	for (int i = 0; i < 9; i++) {
		if (n == grid[r][i]) return false;
	}
	return true;
}

bool checkCol(int c, int n) {
	for (int i = 0; i < 9; i++) {
		if (n == grid[i][c]) return false;
	}
	return true;
}

bool checkSquare(int r, int c, int n) {
	for (int i = 0; i < 3; i++) {
		for (int j = 0; j < 3; j++) {
			if (n == grid[r - r % 3 + i][c - c % 3 + j]) return false;
		}
	}
	return true;
}

void sudoku(int n) {
	if (n == idx) {
		for (int r = 0; r < 9; r++) {
			for (int c = 0; c < 9; c++) {
				cout << grid[r][c] << ' ';
			}
			cout << '\n';
		}
		exit(0);
	}
	for (int i = 1; i <= 9; i++) {
		pii cur = pos[n];
		int r = cur.first;
		int c = cur.second;
		if (checkRow(r, i) && checkCol(c, i) && checkSquare(r, c, i)) {
			grid[r][c] = i;
			sudoku(n + 1);
			grid[r][c] = 0;
		}
	}
}

int main() {
	cin.tie(0)->sync_with_stdio(0);
	
	for (int r = 0; r < 9; r++) {
		for (int c = 0; c < 9; c++) {
			cin >> grid[r][c];
			if (grid[r][c] == 0) pos[idx++] = make_pair(r, c);
		}
	}
	sudoku(0);
}